pkg/clusterd/disk.go

```diff
@@ -55,8 +54,7 @@ func ignoreDevice(d string) bool {
 	return isRBD.MatchString(d)
 }
 
-// DiscoverDevices returns all the details of devices available on the local node
-func DiscoverDevices(executor exec.Executor) ([]*sys.LocalDisk, error) {
+func DiscoverDevicesWithFilter(executor exec.Executor, isFilter bool, deviceName string) ([]*sys.LocalDisk, error) {
```

Comment:
Instead of passing in the individual filter and deviceName properties, we have a double `for` loop, which is not efficient. How about we instead pass in the agent.devices so that inside the existing `for` loop below on line 64, we can decide if the devices should be skipped. Otherwise, we will call `sys.ListDevices()` many times instead of just once.

Reply:
I updated my PR with the suggestion now I'm passing the **agent.devices** instead https://github.com/rook/rook/pull/11976/files#diff-ebe7461a3755ac0d94fec05ffe5a1cee782328dfb6937d237b779a252084ed9fR58  of course we need to keep another loop in order to check the each agent.device since the argument **agent.devices** is an slice of it https://github.com/rook/rook/pull/11976/files#diff-ebe7461a3755ac0d94fec05ffe5a1cee782328dfb6937d237b779a252084ed9fR120 

Although now `sys.ListDevices()` is only called once

Make sense for you?

Reply:
Hi @travisn I noticed with this solution we have and issue with circular dependency

```bash
make check test
make[2]: '/home/javier/src/rook/.cache/tools/go-junit-report' is up to date.
=== go test unit-tests
CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go test -v -cover  -tags '' -pkgdir /home/javier/src/rook/.work/pkg/linux_amd64_static -installsuffix static  -ldflags ' -s -w -X github.com/rook/rook/pkg/version.Version=v1.9.0-alpha.0.1130.g1a0a6fd51-dirty' github.com/rook/rook/cmd/... github.com/rook/rook/pkg/...
package github.com/rook/rook/cmd/rook
	imports github.com/rook/rook/cmd/rook/ceph
	imports github.com/rook/rook/cmd/rook/rook
	imports github.com/rook/rook/pkg/clusterd
	imports github.com/rook/rook/pkg/daemon/ceph/osd
	imports github.com/rook/rook/pkg/operator/ceph/cluster/osd
	imports github.com/rook/rook/pkg/operator/ceph/cluster/mgr
	imports github.com/rook/rook/pkg/operator/ceph/cluster/mon
	imports github.com/rook/rook/pkg/operator/ceph/config
	imports github.com/rook/rook/cmd/rook/rook: import cycle not allowed
package github.com/rook/rook/cmd/rook
	imports github.com/rook/rook/cmd/rook/ceph
	imports github.com/rook/rook/cmd/rook/rook
	imports github.com/rook/rook/pkg/clusterd
	imports github.com/rook/rook/pkg/daemon/ceph/osd
	imports github.com/rook/rook/pkg/clusterd: import cycle not allowed
package github.com/rook/rook/cmd/rook
	imports github.com/rook/rook/cmd/rook/ceph
	imports github.com/rook/rook/cmd/rook/rook
	imports github.com/rook/rook/pkg/clusterd
	imports github.com/rook/rook/pkg/daemon/ceph/osd
	imports github.com/rook/rook/pkg/operator/ceph/cluster/osd
	imports github.com/rook/rook/pkg/operator/ceph/cluster/mgr
	imports github.com/rook/rook/pkg/operator/ceph/cluster/mon
	imports github.com/rook/rook/pkg/operator/ceph/config
	imports github.com/rook/rook/cmd/rook/rook: import cycle not allowed
package github.com/rook/rook/cmd/rook
	imports github.com/rook/rook/cmd/rook/ceph
	imports github.com/rook/rook/cmd/rook/rook
	imports github.com/rook/rook/pkg/clusterd
	imports github.com/rook/rook/pkg/daemon/ceph/osd
	imports github.com/rook/rook/pkg/clusterd: import cycle not allowed
build/makelib/golang.mk:132: recipe for target 'go.test.unit' failed
make[1]: *** [go.test.unit] Error 1
Makefile:146: recipe for target 'check' failed
make: *** [check] Error 2
```
This is happening due this line https://github.com/koor-tech/rook/commit/1a0a6fd51ea6674b0b0fa0e2b60652734ff9bfb3#diff-ebe7461a3755ac0d94fec05ffe5a1cee782328dfb6937d237b779a252084ed9fR58 we need to use `osd.DesiredDevice` inside **clusterd** package.

So I was working on the solution for that and it would be create an interface in **clusterd** package in order to avoid 
circular dependency 

```go
type StorageDevice interface {
	GetName() string
	IsFiltered() bool
	IsDevicePathFiltered() bool
	GetMetadataDevice() string
	SetDeviceClass(deviceClass string)
	GetDeviceClass() string
	MatchWithFilter(deviceName string) bool
}
```
But with this change I need to update a lot of references that are using the struct `DesiredDevice` instead of the interface so I'm a bit concern of this because I'm not sure if that change makes sense